File: src/stores/chronographs.ts
import { invoke } from '@tauri-apps/api/core'
import { createStore } from 'solid-js/store'

export type Chronograph = {
  id: string
  workspace_id: number
  name: string
  kind: 'timer' | 'stopwatch'
  state: 'paused' | 'active'
  duration: number
  is_favourite: boolean
  created_at: string
  modified_at: string
}

export const [chronographs, setChronographs] = createStore<Chronograph[]>([])

export async function fetchAllChronographs(payload: {
  workspace_id: number
  kind: Chronograph['kind']
}) {
  const chronographs = await invoke<Chronograph[]>('fetch_all_chronographs', payload)
  setChronographs(chronographs)
  return chronographs
}

export async function addChronograph(
  chronograph: Omit<Chronograph, 'id' | 'created_at' | 'modified_at'>
) {
  setChronographs((prev) => [
    ...prev,
    { ...chronograph, id: (prev[prev.length - 1]?.id ?? -1) + 1 } as Chronograph,
  ])

  return await invoke<Chronograph>('add_chronograph', { chronograph })
}

export async function updateChronograph(payload: {
  workspace_id: number
  id: string
  chronograph: Partial<Omit<Chronograph, 'id' | 'workspace_id' | 'created_at' | 'modified_at'>>
}) {
  const { id } = payload

  setChronographs((prev) => {
    let index = 0
    let item
    for (index; index < prev.length; index++) {
      item = prev[index]
      if (item.id === id) {
        prev[index] = { ...item, ...payload.chronograph }
      }
    }

    return prev.slice()
  })

  return await invoke<boolean>('update_chronograph', payload)
}

export async function deleteChronograph(payload: { workspace_id: number; id: string }) {
  setChronographs((prev) => prev.filter((chronograph) => chronograph.id !== payload.id))
  return await invoke<boolean>('delete_chronograph', payload)
}

File: tauri/src/commands/chronographs.rs
use serde::{Deserialize, Serialize};
use sqlx::prelude::FromRow;

use crate::db::DatabaseState;

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Chronograph {
    pub id: u32,
    pub workspace_id: u32,
    pub name: String,
    pub kind: String,
    pub state: String,
    pub duration: i32,
    pub is_favourite: bool,
    pub created_at: String,
    pub modified_at: String,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct ChronographInput {
    pub workspace_id: u32,
    pub name: String,
    pub kind: String,
    pub state: String,
    pub duration: i32,
    pub is_favourite: bool,
}

#[tauri::command(rename_all = "snake_case")]
pub async fn fetch_all_chronographs(
    state: tauri::State<'_, DatabaseState>,
    workspace_id: u32,
    kind: String,
) -> Result<Vec<Chronograph>, tauri::Error> {
    let statement = r#"
    SELECT *
    FROM chronographs
    WHERE
        workspace_id = $1
    AND
        kind = $2
    ORDER BY id DESC;
  "#;

    let pool = &state.pool;
    let result = sqlx::query_as::<_, Chronograph>(statement)
        .bind(&workspace_id)
        .bind(&kind)
        .fetch_all(pool)
        .await
        .expect("Fetching all chronographs failed!");

    Ok(result)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn add_chronograph(
    state: tauri::State<'_, DatabaseState>,
    chronograph: ChronographInput,
) -> Result<bool, tauri::Error> {
    let statement = r#"
    INSERT INTO chronographs
    (workspace_id, name, kind, state, duration)
    VALUES ($1, $2, $3, $4, $5);
  "#;

    let pool = &state.pool;
    let result = sqlx::query(statement)
        .bind(&chronograph.workspace_id)
        .bind(&chronograph.name)
        .bind(&chronograph.kind)
        .bind(&chronograph.state)
        .bind(&chronograph.duration)
        .execute(pool)
        .await
        .expect("Adding chronograph failed!");

    if result.rows_affected() == 0 {
        return Ok(false);
    }

    Ok(true)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn update_chronograph(
    state: tauri::State<'_, DatabaseState>,
    chronograph: ChronographInput,
) -> Result<bool, tauri::Error> {
    let statement = r#"
    UPDATE chronographs
    SET workspace_id = $1,
        name = $2,
        kind = $3
        state = $4
        duration = $5
    WHERE id = 0;
  "#;

    let pool = &state.pool;
    let result = sqlx::query(statement)
        .bind(&chronograph.workspace_id)
        .bind(&chronograph.name)
        .bind(&chronograph.kind)
        .bind(&chronograph.state)
        .bind(&chronograph.duration)
        .execute(pool)
        .await
        .expect("Updating chronograph failed!");

    if result.rows_affected() == 0 {
        return Ok(false);
    }

    Ok(true)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn delete_chronograph(
    state: tauri::State<'_, DatabaseState>,
    workspace_id: u32,
    id: u32,
) -> Result<bool, tauri::Error> {
    let statement = r#"
     DELETE FROM chronographs
     WHERE
        workspace_id = $1
     AND
        id = $2;
    "#;

    let pool = &state.pool;
    let result = sqlx::query(statement)
        .bind(&id)
        .bind(&workspace_id)
        .execute(pool)
        .await
        .expect("Deleting chronograph failed!");

    if result.rows_affected() == 0 {
        return Ok(false);
    }

    Ok(true)
}
